### DISCUSSION TITLE
Bloc to Bloc e Clean Arch

### DISCUSSION THREAD

**Discussion by brunosemfio** [Flutterando/forum#115] [Perguntas e Respostas]
Segundo a [documentação](https://bloclibrary.dev/#/architecture?id=bloc-to-bloc-communication) um bloc não deveria depender de outro. Como alternativa, é sugerido que os blocs escutem uma stream no repository.

Na clean architecture o bloc não teria acesso direto ao repository. Então faz sentido criar um usecase apenas para expor uma stream?

Como seria uma abordagem (clean) interessante para evitar dependências dentre blocs?

**Comment by maurojohann** [comment]
Para entender melhor,  voce quer que dois blocs escutem o mesmo repositório, ou que um bloc reaja a outro bloc??

**Reply by brunosemfio** [reply]
Digamos que eu tenha um bloc para autenticação. Quando o usuário fizer login preciso carregar as informações A, B e C, sendo cada informação um bloc diferente. Uma opção seria injetar o bloc de autenticação nos blocs A, B e C. Outra opção, segundo a documentação, é que o `AuthRepository` exponha uma stream do status de autenticação, e que os blocs A, B e C dependam desse repository.

No clean dart, por exemplo, o bloc se comunica com um usecase, que por sua vez se comunica com o repository. Nessa situação, a stream com o status de autenticação deveria ficar no usecase ou, o bloc deveria pular o usecase e se comunicar com o repository?

E por fim, que outro modo os bloc A, B e C poderiam reagir ao bloc de autenticação?

**Reply by maurojohann** [reply]
Voce pode sim fazer um useCase que tenha uma stream, e fazer que todos os blocs dependam dele. E ai quando essa stream emitir um novo status, o bloc chama dentro dele um evento.. Seria algo como assim:
`authUseCase.status.listen( (status){ add(_eventoDoBloc))`..
Mas talvez essa nao seja a melhor opção para verificar um status, fazer que mais de um bloc dependa do authRepository. Talvez em um nivel mais alta da arvore de widget, ter um blocListner que reagisse apenas ao bloc de autenticação.

**Reply by brunosemfio** [reply]
Acho que um BlocListener no topo do app seria uma opção mais interessante. Ai basta emitir um evento em todos os blocs que precisam do status da autenticação.

**Reply by maurojohann** [reply]
Voce está tentando verificar se o usuario ainda está logado pra fazer determinada ação?

**Reply by brunosemfio** [reply]
Sim. Tbm vou precisar fazer algumas verificações antes da ação principal.

Para mais contexto:

O usuário vai reservar X coisa. Antes de processar o pagamento, preciso verificar se está logado, se tem um cartão de crédito adicionado e se está a uma distância válida.

**Reply by maurojohann** [reply]
O bloc da uma opções no exemplo de login dele, acho que vale apena conferir. Mas eu utilizaria um `blocProvider<Authbloc>` em algum lugar que ficasse noma posição ascendente na arvore de widget, e em cada pagina que precisa-se do status teria um blocListner, escutando  Authbloc. Assim antes de qualquer ação dentro dessa pagina, chamaria o evento do Authbloc via context. 
Existem outras opções tbm.
